fog/ingest/enclave/impl/src/lib.rs

```diff
@@ -383,3 +389,45 @@ impl<OSC: ORAMStorageCreator<StorageDataSize, StorageMetaSize>> IngestEnclave
 fn seal_private_key(src: &RistrettoPrivate) -> Result<SealedIngestKey> {
     Ok(IntelSealed::seal_raw(src.as_ref(), &[])?.as_ref().to_vec())
 }
+
+// Helper to work around TxOut => FogTxOut being fallible.
+// During attempt_ingest_txs, we must be constant-time and it complicates things
+// if operations can fail, because an adversary may be able to use that as
+// leverage.
+//
+// Instead we want to convert all TxOut to FogTxOut up front before calling
+// attempt_ingest_txs, so that all the failures happen before anything sensitive
+// happens.
+#[derive(Clone, Eq, PartialEq, Debug)]
+struct PreparedBlockData {
+    /// The index of the block in the blockchain where this chunk of Txs
+    /// appeared
+    pub block_index: u64,
+    /// The number of txo's appearing in the entire blockchain before this chunk
+    /// This is needed to compute TxOutRecord::global_tx_out_index
+    pub global_txo_index: u64,
+    /// The txout's of this chunk, also formatted as FogTxOut
+    pub tx_outs: Vec<(TxOut, FogTxOut)>,
+    /// The timestamp of this block
+    pub timestamp: u64,
+}
+
+impl TryFrom<TxsForIngest> for PreparedBlockData {
+    type Error = Error;
+
+    fn try_from(src: TxsForIngest) -> Result<PreparedBlockData> {
+        Ok(PreparedBlockData {
+            block_index: src.block_index,
+            global_txo_index: src.global_txo_index,
+            timestamp: src.timestamp,
+            tx_outs: src
+                .redacted_txs
+                .into_iter()
+                .map(|txo| match FogTxOut::try_from(&txo) {
+                    Ok(ftxo) => Ok((txo, ftxo)),
+                    Err(_) => Err(Error::InvalidBlockData),
```

Comment:
i think right now it's like, this would require `AmountError` to support `serde` because these errors are serde serialized. but there are separate efforts going on to make a bunch of transaction core stuff be no-alloc friendly to support hardware wallets. so i'm worried about adding more serde here and creating more work here.

maybe we should just turn the error to string to sidestep the serde issue?

Reply:
i guess i shold just make those errors serde as well, there is a lot of serde stuff in transaction core and whatever solution is used for all of that should apply just fine for these errors too